modules/salt/pages/salt-rate-limiting.adoc

```diff
@@ -0,0 +1,54 @@
+[[salt.rate.limiting]]
+= Rate Limiting
+
+Salt is able to run many commands in parallel on a large number of clients.
+This can potentially create large amounts of load on your infrastructure.
+You can use these rate-limiting features to control the load in your environment.
+
+These features are all configured in the [filename]``/etc/rhn/rhn.conf`` configuration file.
+
+
+
+== Batch Size
+
+Actions are processed in batches.
+This parameter sets the maximum number of clients that can execute a single action in each batch.
```

Comment:
Not really.

Batching and rate limiting are different concepts, and what Salt/`{productname}` does is rate limiting, not batching. It's unfortunate that this very feature is named Batching in Salt (upstream) and existed long before we added support in `{productname}`, thus we had to adopt its jargon.

Example: a command has to be executed on 100 minions, and the batch size is 20.

**Batching** means:
 1. that 5 sets of 20 minions each are created
 1. that initially, 20 commands are issued (one per minion in batch 1)
 1. that as soon as all 20 minions in batch 1 are done, batch 2 begins and so on
 1. thus, that at any point in time there can only be 20 or less commands in execution
 1. typically the number will be 20 when batches are started and will gradually diminish until batch end. On average, less than 20 will be active

**Rate limiting** means:
 1. that only 1 set of 20 minions is created (**unlike** in batching)
 1. that initially, 20 commands are issued (one per minion in the only one batch) (**same** as in batching)
 1. that as soon as any minion is done, another minion immediately* starts (**unlike** in batching)
    - corner case: more than one minion is done precisely at the same time. An equal number of minions start
 1. thus, that at any point in time there can only be 20 or less commands in execution (**same** as in batching)
 1. typically the number will be exactly 20, give or take a few during issuing. On average, the number of active commands will be very close to 20 (**unlike** in batching)

*Note the asterisk at point 3. More about that later*

Summarizing: both batching and rate limiting guarantee no more than 20 commands are active at any given point in time, but rate limiting does a much better job in keeping the number close to 20 for the whole duration, resulting in better performance. That's why the latter is implemented and not the former!


Hopefully the SUSECON presentation I sent you (in particular the animations) make this clearer.

Reply:
Yes, the presentation (and your animations!) were very helpful. I wasn't aware of this specific difference, though.